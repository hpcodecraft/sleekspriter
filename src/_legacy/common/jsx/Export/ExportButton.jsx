var ExportButton = React.createClass({
  mixins: [FluxMixin, TouchMixin],
  render: function() {
    return (
      <a className="button" ref="theButton" onClick={this.handleClick.bind(this, this.export)} onTouchStart={this.handleTouch.bind(this, this.export)}>Export</a>
    )
  },
  export: function() {
    platformUtils.exportFile(this.props);
  },
});
