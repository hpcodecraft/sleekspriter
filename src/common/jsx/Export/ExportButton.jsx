var ExportButton = React.createClass({
  mixins: [FluxMixin],
  render: function() {
    return (
      <a className="button" ref="theButton" onClick={this.export}>Export</a>
    )
  },
  export: function() {
    platformUtils.exportFile(this.props);
  },
});