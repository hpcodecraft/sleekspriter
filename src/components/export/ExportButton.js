import React from "react";
import { t } from "../../utils";

const ExportButton = (props) => {
  return ( <a className="button" onClick={props.export}>{t("Export")}</a> );
};

export default ExportButton;
