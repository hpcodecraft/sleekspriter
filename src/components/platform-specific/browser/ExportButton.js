import React from "react";
import PropTypes from "prop-types";

import { t } from "../../../utils";
import config from "../../../config";
import { exportFormats } from "../../../const";

const ExportButton = props => {
  const exportFile = () => {
    const { file, format, exportStatus } = props;

    const fileName = file.name || config.defaultName;

    const mimeType =
      format === exportFormats.jpg ? "image/jpeg" : `image/${format}`;

    const canvas = document
      .getElementById("ExportPreview")
      .querySelectorAll("canvas")[0];
    const img = canvas.toDataURL(mimeType, 1);
    const downloadLink = document.createElement("a");

    downloadLink.href = img;
    downloadLink.download = `${fileName}.${format}`;

    document.body.appendChild(downloadLink);
    downloadLink.click();
    document.body.removeChild(downloadLink);

    exportStatus(t("Export finished"));

    setTimeout(() => {
      exportStatus("");
    }, 5000);
  };

  return (
    <a
      className="button"
      onClick={() => {
        exportFile(props);
      }}>
      {t("Export")}
    </a>
  );
};

ExportButton.propTypes = {
  file: PropTypes.object.isRequired,
  format: PropTypes.string.isRequired,
  exportStatus: PropTypes.func.isRequired,
};

export default ExportButton;
