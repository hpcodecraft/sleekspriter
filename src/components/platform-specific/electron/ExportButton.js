import React from "react";
import PropTypes from "prop-types";
import { remote } from "electron";
import { writeFile } from "fs";

import { exportFormats } from "../../../const";
import { t } from "../../../utils";

import { fileDefaultPath } from "./utils";

const { dialog } = remote;

// Decoding base-64 image
// Source: http://stackoverflow.com/questions/20267939/nodejs-write-base64-image-file
function decodeBase64Image(dataString) {
  var matches = dataString.match(/^data:([A-Za-z-+/]+);base64,(.+)$/);
  var response = {};

  if (matches.length !== 3) {
    return new Error("Invalid input string");
  }

  response.type = matches[1];
  response.data = Buffer.from(matches[2], "base64");

  return response;
}

const ExportButton = props => {
  const handleError = error => {
    console.error(error);
  };

  const exportFile = () => {
    const { exportStatus, file, format, zoom } = props;

    const defaultPath = fileDefaultPath(file, format, zoom);

    const mimeType =
      format === exportFormats.jpg ? "image/jpeg" : `image/${format}`;

    const canvas = document
      .getElementById("ExportPreview")
      .querySelectorAll("canvas")[0];

    const filePath = dialog.showSaveDialogSync({
      defaultPath,
      filters: [{ name: "Image", extensions: [format] }],
    });

    if (filePath) {
      const img = canvas.toDataURL(mimeType, 1);
      const buffer = decodeBase64Image(img);

      writeFile(filePath, buffer.data, function(error) {
        if (error) {
          handleError(error);
        }

        exportStatus(t("Export finished"));

        setTimeout(() => {
          exportStatus("");
        }, 5000);
      });
    }
  };

  return (
    <a
      className="button"
      onClick={() => {
        exportFile(props);
      }}>
      {t("Export")}
    </a>
  );
};

ExportButton.propTypes = {
  exportStatus: PropTypes.func.isRequired,
  file: PropTypes.object.isRequired,
  format: PropTypes.string.isRequired,
  zoom: PropTypes.number.isRequired,
};

export default ExportButton;
