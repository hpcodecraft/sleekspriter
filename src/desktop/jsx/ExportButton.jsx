var ExportButton = React.createClass({
  mixins: [FluxMixin],
  render: function() {
    return (
      <a className="button" ref="theButton" onClick={this.export}>Export</a>
    )
  },
  export: function() {
    var self = this,
        fs = require('fs'),
        sys = require('sys'),
        path = require('path'),
        finishedStatusText;

    if(this.props.file.folder === null) {
      this.getFlux().actions.modalShow(ModalErrorSaveBeforeExport);
      return;
    }

    switch(this.props.ui.export.part) {
      case 'animation':
        var animation = storeUtils.animations.getById(this.props.ui.export.animation);
        finishedStatusText = 'Exported to '+this.props.file.folder+path.sep+this.props.file.name+'-'+animation.name+'.gif';
        writeAnimation();
        break;

      case 'allframes':
        finishedStatusText = 'Exported '+self.props.ui.frames.total+' frames as '+(this.props.ui.export.format === 'jpeg' ? 'jpg' : this.props.ui.export.format)+' to '+this.props.file.folder;
        writeImages();
        break;

      default:
        finishedStatusText = 'Exported to '+this.props.file.folder+path.sep+this.props.file.name+'.'+(this.props.ui.export.format === 'jpeg' ? 'jpg' : this.props.ui.export.format);
        writeImages();
        break;
    }

    function writeAnimation() {
      var gif = new GIF({
                      workers: 2,
                      quality: 1,
                      // background: '#ff0000',
                      transparent: 0x000000,
                    }),
          canvas = NodeList2Array(document.getElementById('ExportPreview').querySelector('.animation-frames').querySelectorAll('.preview')),
          delay = 1000/animation.fps;

      canvas.forEach(function(canvas) {
        gif.addFrame(canvas, {delay: delay});
      });

      gif.on('finished', exportGif);
      gif.render();

      function exportGif(blob) {
       var reader = new FileReader();
       reader.onload = writeGif;
       reader.readAsDataURL(blob);
      }

      function writeGif(e) {
        var data = e.target.result.replace(/^data:image\/\w+;base64,/, ""),
            buf = new Buffer(data, 'base64'),
            target = self.props.file.folder+'/'+self.props.file.name+'-'+self.props.ui.export.animation+'.gif';

        fs.writeFile(target, buf);

        self.getFlux().actions.exportStatus(finishedStatusText);
      }
    }


    function writeImages() {
      var canvas = NodeList2Array(document.getElementById('ExportPreview').querySelectorAll('.preview')),
          frame = 1;

      function saveCanvas(canvas, fileName) {
        var img = canvas.toDataURL('image/'+self.props.ui.export.format),
            data = img.replace(/^data:image\/\w+;base64,/, ""),
            buf = new Buffer(data, 'base64'),
            target = self.props.file.folder+'/'+fileName+'.'+(self.props.ui.export.format === 'jpeg' ? 'jpg' : self.props.ui.export.format);

        fs.writeFile(target, buf);
      }

      canvas.forEach(function(canvas) {
        var fileName = self.props.file.name;
        if(self.props.ui.export.part === 'oneframe') fileName+= '-frame-'+self.props.ui.export.frame;
        else if(self.props.ui.export.part === 'allframes') {
          fileName+= '-frame-'+frame;
          frame++;
        }
        saveCanvas(canvas, fileName);
      });

      self.getFlux().actions.exportStatus(finishedStatusText);
    }

  },
});