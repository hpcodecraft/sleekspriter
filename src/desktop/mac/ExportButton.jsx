var ExportButton = React.createClass({
  render: function() {
    return (
      <a className="button" ref="theButton" onClick={this.export}>Export</a>
    )
  },
  export: function() {
    var self = this,
        fs = require('fs'),
        sys = require('sys');

    if(this.props.part === 'animation') {
      var gif = new GIF({
                      workers: 2,
                      quality: 1,
                      // background: '#ff0000',
                      transparent: 0x000000,
                    }),
          canvas = NodeList2Array(document.getElementById('ExportPreview').querySelector('.animation-frames').querySelectorAll('.preview')),
          delay = 1000/this.props.editor.animations.getByName(this.props.animation).fps;

      function exportGif(blob) {
       var reader = new FileReader();
       reader.onload = writeGif;
       reader.readAsDataURL(blob);
      }

      function writeGif(e) {
        var data = e.target.result.replace(/^data:image\/\w+;base64,/, ""),
            buf = new Buffer(data, 'base64'),
            target = file.folder+'/'+file.name+'-'+self.props.animation+'.gif';

        fs.writeFile(target, buf);

        channel.publish('export.finished', {
          folder: file.folder,
          name: file.name,
          format: self.props.format,
          part: self.props.part,
          frames: self.props.editor.frames.total,
          animation: self.props.animation,
        });
      }

      canvas.forEach(function(canvas) {
        gif.addFrame(canvas, {delay: delay});
      });

      gif.on('finished', exportGif);
      gif.render();
    }
    else {
      var canvas = NodeList2Array(document.getElementById('ExportPreview').querySelectorAll('.preview')),
          frame = 1;

      function saveCanvas(canvas, fileName) {
        var img = canvas.toDataURL('image/'+self.props.format),
            data = img.replace(/^data:image\/\w+;base64,/, ""),
            buf = new Buffer(data, 'base64'),
            target = file.folder+'/'+fileName+'.'+(self.props.format === 'jpeg' ? 'jpg' : self.props.format);

        fs.writeFile(target, buf);
      }

      canvas.forEach(function(canvas) {
        var fileName = file.name;
        if(self.props.part === 'oneframe') fileName+= '-frame-'+self.props.frame;
        else if(self.props.part === 'allframes') {
          fileName+= '-frame-'+frame;
          frame++;
        }
        saveCanvas(canvas, fileName);
      });

      channel.publish('export.finished', {
        folder: file.folder,
        name: file.name,
        format: this.props.format,
        part: this.props.part,
        frames: this.props.editor.frames.total,
      });
    }
  },
});