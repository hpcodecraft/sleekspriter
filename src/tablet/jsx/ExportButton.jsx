var ExportButton = React.createClass({
  render: function() {
    return (
      <a className="button" ref="theButton" onClick={this.export}>Export</a>
    )
  },
  export: function() {
    var canvas = NodeList2Array(document.getElementById('ExportPreview').querySelectorAll('.preview'));

    if(canvas.length === 1) {
      var fileName = file.name,
          node = this.refs.theButton.getDOMNode();

      if(this.props.part === 'oneframe') fileName+= '-frame-'+this.props.frame;

      node.download = fileName+'.'+(this.props.format === 'jpeg' ? 'jpg' : this.props.format);
      node.href = canvas[0].toDataURL('image/'+this.props.format);
      channel.gui.publish('export.finished', {folder: file.folder, name: fileName, format: this.props.format, part: this.props.part});
    }
  },
});