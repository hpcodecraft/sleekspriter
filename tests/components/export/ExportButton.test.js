import ExportButton from "components/export/ExportButton";

describe("ExportButton", () => {
  let props, wrapper;
  beforeEach(() => {
    props = {
      export: jest.fn(),
    };
    wrapper = shallow(<ExportButton {...props} />);
  });

  it("should render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("fires export handler on click", () => {
    wrapper.simulate("click");
    expect(props.export).toHaveBeenCalledTimes(1);
  });
});
