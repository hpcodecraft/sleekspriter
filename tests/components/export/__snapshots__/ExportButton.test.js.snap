// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportButton should render correctly 1`] = `"<a class=\\"button\\">Export</a>"`;
